### PR TITLE
improve player NPC skins

### PR DESCRIPTION
**Pull request by JCThePants**
add skin packages with skin classes

add profile package with profile fetcher classes

update commands

update EventListen

update NMS

### CODE REVIEW

**Comment by JCThePants** [comment]
"fix" for issue #468

My test build is available at https://www.dropbox.com/s/9lvd9p8gw8gdozh/CitizensTest1.8.8.jar?dl=0

**Comment by JCThePants** [comment]
Test build in drop box link updated

**Comment by JCThePants** [comment]
Test build in drop box link updated

File: src/main/java/net/citizensnpcs/EventListen.java [hunk 1]
```diff
@@ -423,34 +426,66 @@ public void onWorldUnload(WorldUnloadEvent event) {
...
+        for (NPC npc : getAllNPCs()) {
+
+            Entity npcEntity = npc.getEntity();
+            if (npcEntity instanceof Player
+                    && player.canSee((Player) npcEntity)
+                    && player.getWorld().equals(npcEntity.getWorld())
+                    && player.getLocation().distanceSquared(npcEntity.getLocation()) < 100 * 100) {
+
+                EntitySkinnable skinnable = NMS.getSkinnableNPC(npcEntity);
+                assert skinnable != null;
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: Asserts are disabled on most JVMs. Don't think this is necessary in the release build.

File: src/main/java/net/citizensnpcs/EventListen.java [hunk 2]
```diff
@@ -423,34 +426,66 @@ public void onWorldUnload(WorldUnloadEvent event) {
...
+
+        List<EntitySkinnable> results = new ArrayList<EntitySkinnable>();
 
+        for (NPC npc : getAllNPCs()) {
+
+            Entity npcEntity = npc.getEntity();
+            if (npcEntity instanceof Player
+                    && player.canSee((Player) npcEntity)
+                    && player.getWorld().equals(npcEntity.getWorld())
+                    && player.getLocation().distanceSquared(npcEntity.getLocation()) < 100 * 100) {
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: I'd prefer if you could use the cached Location methods (npc.getStoredLocation() and player.getLocation(Location)) to avoid object churn.

Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: Also, the 100 * 100 number should be a named static constant.

File: src/main/java/net/citizensnpcs/EventListen.java [hunk 3]
```diff
@@ -559,4 +570,45 @@ public int hashCode() {
...
+
+            if (!hasMoved) {
+                hasMoved = true;
+                return true;
+            }
+
+            Location currentLoc = player.getLocation(YAW_LOCATION);
+            float currentYaw = currentLoc.getYaw();
+
+            boolean hasRotated = Math.abs(currentYaw - this.initialYaw) < 90;
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: I think Minecraft yaws need to be changed using two while loops to put them in the right range. You may end up with bugs if you use it like this directly. Also, is this just for small rotations? Not sure why 90 is being used here.

while(yaw < 180) yaw += 180;
while(yaw > 180) yaw -= 180;

File: src/main/java/net/citizensnpcs/EventListen.java [hunk 4]
```diff
@@ -559,4 +570,45 @@ public int hashCode() {
...
+
+            // update the first 2 times the player rotates. helps load skins around player
+            // after the player logs/teleports.
+            if (hasRotated && rotationCount < 2) {
+                rotationCount++;
+                return true;
+            }
+
+            double distance = currentLoc.distanceSquared(this.location);
+            return distance > 50 * 50;
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: Make into named static constant.

File: src/main/java/net/citizensnpcs/npc/entity/HumanController.java [hunk 5]
```diff
@@ -1,57 +1,62 @@
...
+    }
+
+    @Override
+    protected boolean couldSpawn(Location at, NPC npc) {
+        if (!super.couldSpawn(at, npc)) {
+            return false;
         }
+
+        EntitySkinnable skinnable = NMS.getSkinnableNPC(npc.getEntity());
+        assert skinnable != null;
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/entity/HumanController.java]: delete asserts

File: src/main/java/net/citizensnpcs/npc/entity/HumanController.java [hunk 6]
```diff
@@ -1,57 +1,62 @@
...
+
+        skinnable.getSkinTracker().updateNearbyViewers(100.0);
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/entity/HumanController.java]: 100 constant seems arbitrary but used throughout, maybe need to merge into somewhere (possibly a setting).

File: src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            request = new ProfileRequest(name, subscriber);
+            queue.add(request);
+            requested.put(name, request);
+        }
+    }
+
+    @Override
+    public void run() {
+
+        if (Bukkit.isPrimaryThread())
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: this calls Thread.currentThread(), which can slow down execution a bit. I would avoid calling it more than once personally.

File: src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+        synchronized (sync) {
+
+            ProfileRequest request = requested.get(name);
+
+            if (request != null) {
+
+                if (subscriber != null) {
+
+                    if (request.getResult() == ProfileFetchResult.PENDING) {
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: I'm not 100% sure, but I think this might cause async issues sometimes since nothing is being synced here. If the result is set right after this if block is intered but before request.addSubscriber is called then subscriber.onResult might not be called.

File: src/main/java/net/citizensnpcs/npc/skin/Skin.java [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+     * @param entity  The human NPC entity.
+     *
+     * @return  True if the skin data was available and applied, false if
+     * the data is being retrieved.
+     *
+     * @throws IllegalStateException if not invoked from the main thread.
+     */
+    public boolean apply(EntitySkinnable entity) {
+        Preconditions.checkNotNull(entity);
+        assertMainThread();
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/skin/Skin.java]: again, asserting main thread is slow.

File: src/main/java/net/citizensnpcs/npc/skin/EntitySkinnable.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
+package net.citizensnpcs.npc.skin;
+
+import com.mojang.authlib.GameProfile;
+import net.citizensnpcs.npc.ai.NPCHolder;
+import org.bukkit.entity.Player;
+
+/**
+ * Interface for player entities that are skinnable.
+ */
+public interface EntitySkinnable extends NPCHolder {
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/skin/EntitySkinnable.java]: Would prefer 'SkinnableEntity' to keep with the Bukkit naming convention.

File: src/main/java/net/citizensnpcs/npc/CitizensNPC.java [hunk 11]
```diff
@@ -185,10 +180,9 @@ public boolean spawn(Location at) {
         at = at.clone();
         getTrait(CurrentLocation.class).setLocation(at);
 
-        entityController.spawn(at, this);
+        boolean couldSpawn = entityController.spawn(at, this);
 
         net.minecraft.server.v1_8_R3.Entity mcEntity = ((CraftEntity) getEntity()).getHandle();
-        boolean couldSpawn = !Util.isLoaded(at) ? false : mcEntity.world.addEntity(mcEntity, SpawnReason.CUSTOM);
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/CitizensNPC.java]: Why was this changed?

**Comment by fullwall** [comment]
Hi, really great job at understanding the current approach and then making a fix for it, lot of good code here. I need to understand it a bit more for maintainability purposes before I can merge it I think, I didn't expect so much extra async code being added - I've made some comments. I'm worried that there might be an additional abstraction over skin subscribers etc than necessary.

File: src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+                        request.addSubscriber(subscriber);
+                    }
+                    else {
+                        subscriber.onResult(request);
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: This is guaranteed to be off the main thread as well.

File: src/main/java/net/citizensnpcs/npc/profile/ProfileRequest.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+     *
+     * <p>Subscriber is always invoked from the main thread.</p>
+     *
+     * @param subscriber  The subscriber.
+     */
+    public void addSubscriber(ProfileFetchSubscriber subscriber) {
+        Preconditions.checkNotNull(subscriber);
+
+        if (subscribers == null)
+            subscribers = new LinkedList<ProfileFetchSubscriber>();
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileRequest.java]: Would prefer ArrayDeque

File: src/main/java/net/citizensnpcs/npc/skin/EntitySkinnable.java (hunk 10 above)
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/skin/EntitySkinnable.java]: Could also put this onto the CraftEntity portion of the EntityPlayer

File: src/main/java/net/citizensnpcs/EventListen.java (hunk 3 above)
Comment (JCThePants) [inline src/main/java/net/citizensnpcs/EventListen.java]: Oh wow, completely forgot about that.

I'll explain the turning issue. In testing, if I log in or am teleported, if I wait perhaps 5 seconds or so before turning around, NPCs behind me do not load skins. I'm sending updates after the player turns at least 90 degrees. I experimented with smaller numbers than 90 but it caused more updates and more tab list flashing and didn't seem to provide any benefit over 90.

Ill add a comment.

File: src/main/java/net/citizensnpcs/npc/CitizensNPC.java (hunk 11 above)
Comment (JCThePants) [inline src/main/java/net/citizensnpcs/npc/CitizensNPC.java]: I was trying to send the playerlist packet after the entity was added to the world but couldn't do that easily from the controller spawn method since it happened. TBH, I'm not 100% sure if it even matters. I can try putting it back and doing more tests.

**Comment by JCThePants** [comment]
@fullwall I was wondering if you prefer a single commit to fix these or a commit per issue

File: src/main/java/net/citizensnpcs/npc/CitizensNPC.java (hunk 11 above)
Comment (JCThePants) [inline src/main/java/net/citizensnpcs/npc/CitizensNPC.java]: to clarify, I made that part a protected method so I could override it in the human controller.

File: src/main/java/net/citizensnpcs/EventListen.java (hunk 3 above)
Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: Yes, I understand the turning issue because that's why skins don't load most of the time (I verified in MC source a while ago). I think this is where you get most of the benefits.

Comment (fullwall) [inline src/main/java/net/citizensnpcs/EventListen.java]: The larger the constant is the more likely it is to update right? That's the way I read it.

Comment (JCThePants) [inline src/main/java/net/citizensnpcs/EventListen.java]: Correct. If it happened too soon in the turn it sometimes didn't work.

File: src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java (hunk 8 above)
Comment (JCThePants) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: I added the sync to synchronize the queue. The request result is always set on the main thread no matter which thread it's invoked from.

I used a sync object instead of of blocking queue so that i could sync multiple operations on the queue. I was worried that, for instance, checking if the queue is empty might not return the same result after executing. Correct me if I'm wrong on this.

Looking at it though, I am syncing way more code than I need to.

File: src/main/java/net/citizensnpcs/npc/entity/HumanController.java (hunk 6 above)
Comment (JCThePants) [inline src/main/java/net/citizensnpcs/npc/entity/HumanController.java]: No problem. I only used 100 because it's what was there already.

**Comment by JCThePants** [comment]
Test build in drop box link updated

**Comment by mcmonkey4eva** [comment]
@JCThePants You see that it's unmergeable atm, right?
Should probably fix that + squash a little before continuing.

**Comment by JCThePants** [comment]
@mcmonkey4eva Appreciate the tip!

**Comment by JCThePants** [comment]
Having issues with NPC's occasionally not being visible at all. Looking into it.

**Comment by fullwall** [comment]
Not being visible at all sounds like you're not sending the tablist packet early enough before the spawn packet.

File: src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Thread used to fetch profiles from the Mojang servers.
+ *
+ * <p>Maintains a cache of profiles so that no profile is ever requested more than once
+ * during a single server session.</p>
+ */
+public class ProfileFetchThread implements Runnable {
+
+    private final ProfileFetcher profileFetcher = new ProfileFetcher();
+    private final Deque<ProfileRequest> queue = new LinkedList<ProfileRequest>();
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: ArrayDeque?

File: src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    public void fetch(String name, @Nullable ProfileFetchSubscriber subscriber) {
+        Preconditions.checkNotNull(name);
+
+        ProfileRequest request = requested.get(name);
+
+        if (request != null) {
+
+            if (subscriber != null) {
+
+                if (request.getResult() == ProfileFetchResult.PENDING) {
```
Comment (fullwall) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: Still not sure if this is correct in terms of async. If setResult is performed on a different thread to this thread I think it's not async safe.

Comment (JCThePants) [inline src/main/java/net/citizensnpcs/npc/profile/ProfileFetchThread.java]: Gotcha. I was still thinking in terms of you cant call fetch from any thread other than main.

**Comment by JCThePants** [comment]
I managed to fix the dissappearing NPC's last night, just waiting for input before I pushed more changes. Not exactly sure but it seemed that sometimes I had to send the packet more than once to ensure the client registered it. That may not be correct at all but allowing new packet tasks/sets to cancel the current one instead of blocking new requests until the current one finishes has fixed it.

**Comment by JCThePants** [comment]
Oops.

**Comment by JCThePants** [comment]
Test build in drop box link updated
